### PR TITLE
worker/compute: Retry local dispatch between workers

### LINKED ISSUE #3000: Local dispatch between workers should be retried
Currently there are special cases when the local node has multiple roles simultaneously:
* [txn scheduler and executor](https://github.com/oasisprotocol/oasis-core/blob/3ffcef81af553ce198de5d08188c58be79ed98f2/go/worker/compute/executor/committee/node.go#L296)
* [executor and merge](https://github.com/oasisprotocol/oasis-core/blob/3ffcef81af553ce198de5d08188c58be79ed98f2/go/worker/compute/merge/committee/node.go#L275)

In these cases the local submission does not go via P2P but via a direct function call. However such dispatch is never retried in case of failures (the failure cases are the same as with P2P dispatch). The context that should be used in both cases is the [`roundCtx`](https://github.com/oasisprotocol/oasis-core/blob/3ffcef81af553ce198de5d08188c58be79ed98f2/go/worker/common/committee/group.go#L46) specified in the local `Group` (currently not exposed).

### PR DESCRIPTION
**Pull request by abukosek**
Fixes #3000.

### CODE REVIEW

File: go/worker/common/committee/group.go [hunk 1]
```diff
@@ -255,6 +255,11 @@ type Group struct {
 	logger *logging.Logger
 }
 
+// RoundContext returns the active epoch's round context.
+func (g *Group) RoundContext() context.Context {
+	return g.activeEpoch.roundCtx
```
Comment (kostko) [inline go/worker/common/committee/group.go:260]: Access to `g.activeEpoch` requires the read lock to be held. Also what happens in case `g.activeEpoch` is `nil`?

File: go/worker/compute/executor/committee/node.go [hunk 2]
```diff
@@ -511,7 +518,22 @@ func (n *Node) startProcessingBatchLocked(
 			batchRuntimeProcessingTime.With(n.getMetricLabels()).Observe(time.Since(rtStartTime).Seconds())
 		}()
 
-		rsp, err := rt.Call(ctx, rq)
+		roundCtx := n.commonNode.Group.RoundContext()
+		var rsp *protocol.Body
+		call := func() error {
+			var grr error
+			rsp, grr = rt.Call(ctx, rq)
```
Comment (kostko) [inline go/worker/compute/executor/committee/node.go:525]: Why did you add a retry here? This should not be needed.

The only methods that need retry are (as mentioned in #3000 with links):
* `HandleBatchFromTransactionSchedulerLocked` in executor.
* `HandleResultsFromExecutorWorkerLocked` in merge.

File: go/worker/common/committee/group.go (hunk 1 above)
Comment (abukosek) [reply go/worker/common/committee/group.go:260]: Oops, will add the lock.  Where this is called from, the `activeEpoch` shouldn't be `nil`, though.

File: go/worker/compute/executor/committee/node.go (hunk 2 above)
Comment (abukosek) [reply go/worker/compute/executor/committee/node.go:525]: `HandleBatchFromTransactionSchedulerLocked` calls `maybeStartProcessingBatchLocked`, which calls `startProcessingBatchLocked`, which can fail only in this `rt.Call`.  There doesn't seem to be any other place to put the retry but here.

Comment (kostko) [reply go/worker/compute/executor/committee/node.go:525]: Retrying here does not make sense as the `Call` internally includes a retry in case the runtime is not available. If it can't fail elsewhere, maybe retry is not needed for the transaction scheduler case then? I'll take another look.

Comment (kostko) [reply go/worker/compute/executor/committee/node.go:525]: Yeah it seems that `HandleBatchFromTransactionSchedulerLocked` does not need retry.

Comment (abukosek) [reply go/worker/compute/executor/committee/node.go:525]: OK, I'll remove it then :)

File: go/worker/common/committee/group.go [hunk 3]
```diff
@@ -255,6 +255,16 @@ type Group struct {
 	logger *logging.Logger
 }
 
+// RoundContext returns the active epoch's round context.
+func (g *Group) RoundContext() (*context.Context, error) {
+	g.RLock()
+	defer g.RUnlock()
+	if g.activeEpoch == nil {
```
Comment (kostko) [inline go/worker/common/committee/group.go:262]: Empty line above.

File: go/worker/compute/merge/committee/node.go [hunk 4]
```diff
@@ -76,6 +77,12 @@ var (
 	infiniteTimeout = time.Duration(math.MaxInt64)
 )
 
+// Parameters for retrying local submission calls.
+const (
+	retryInterval = 1 * time.Second
+	maxRetries    = 15
```
Comment (kostko) [inline go/worker/compute/merge/committee/node.go:83]: This should be [the same as in P2P](https://github.com/oasisprotocol/oasis-core/blob/cfc8b1a0d12e0e1fe425d987b6149643b844a015/go/worker/common/p2p/p2p.go#L194) which just uses `NewExponentialBackOff`. No need to limit the retries as the round context will be cancelled in case the round advances.

File: go/worker/common/committee/group.go [hunk 5]
```diff
@@ -255,6 +255,16 @@ type Group struct {
 	logger *logging.Logger
 }
 
+// RoundContext returns the active epoch's round context.
+func (g *Group) RoundContext() (*context.Context, error) {
```
Comment (kostko) [inline go/worker/common/committee/group.go:259]: `context.Context` is an interface and so it can be nil, no need to return a pointer to it.

File: go/worker/compute/merge/committee/node.go [hunk 6]
```diff
@@ -273,8 +280,28 @@ func (n *Node) HandleNewBlockLocked(blk *block.Block) {
...
+		// which shouldn't happen in our case.
+		n.logger.Warn("HandleResultsFromExecutorWorkerLocked called with no active epoch")
+		return
+	}
+
+	call := func() error {
+		grr := n.handleResultsLocked(*ctx, commit)
+		switch grr {
+		case errIncorrectState:
+			return backoff.Permanent(grr)
```
Comment (kostko) [inline go/worker/compute/merge/committee/node.go:295]: Why no retry in this case? The P2P code path retries even if the state is incorrect.

Comment (abukosek) [reply go/worker/compute/merge/committee/node.go]: Fixed :)

File: go/worker/compute/merge/committee/node.go [hunk 7]
```diff
@@ -273,8 +274,21 @@ func (n *Node) HandleNewBlockLocked(blk *block.Block) {
 // HandleResultsFromExecutorWorkerLocked processes results from an executor worker.
 // Guarded by n.commonNode.CrossNode.
 func (n *Node) HandleResultsFromExecutorWorkerLocked(spanCtx opentracing.SpanContext, commit *commitment.ExecutorCommitment) {
-	// TODO: Context.
-	if err := n.handleResultsLocked(context.TODO(), commit); err != nil {
+	ctx, err := n.commonNode.Group.RoundContext()
```
Comment (kostko) [inline go/worker/compute/merge/committee/node.go:277]: Actually I just noticed that there already eixsts a `n.roundCtx` that you can use, so this is not needed.

Comment (abukosek) [reply go/worker/compute/merge/committee/node.go:277]: Haha, this PR is turning into a 2-line fix :D

File: go/worker/compute/merge/committee/node.go [hunk 8]
```diff
@@ -273,8 +274,9 @@ func (n *Node) HandleNewBlockLocked(blk *block.Block) {
 // HandleResultsFromExecutorWorkerLocked processes results from an executor worker.
 // Guarded by n.commonNode.CrossNode.
 func (n *Node) HandleResultsFromExecutorWorkerLocked(spanCtx opentracing.SpanContext, commit *commitment.ExecutorCommitment) {
-	// TODO: Context.
-	if err := n.handleResultsLocked(context.TODO(), commit); err != nil {
+	call := func() error { return n.handleResultsLocked(n.roundCtx, commit) }
+	bkoff := backoff.WithContext(backoff.NewExponentialBackOff(), n.roundCtx)
+	if err := backoff.Retry(call, bkoff); err != nil {
```
Comment (kostko) [inline go/worker/compute/merge/committee/node.go:279]: This probably needs to be changed to perform retry in an async fashion, otherwise this can block all processing and the error condition will never resolve (as the lock is held all the time. So probably just spawn retry in a separate go routine and make sure to take the lock in the call function each time.

File: go/worker/compute/merge/committee/node.go (hunk 7 above)
Comment (kostko) [reply go/worker/compute/merge/committee/node.go]: :-)
